### ISSUE TITLE
"Annotation from val declaration overrides inline type annotation" warnings are too eager

### ISSUE THREAD

**Issue by cpitclaudel** [FStarLang/FStar#1157]
The following code:

```
module Min

type a = | A
val _a : a
let _a : a = A
```

produces a warning (Annotation from val declaration overrides inline type annotation).  Should it really?

**Comment by kyoDralliam** [comment]
The compiler just generates this warning whenever it encounters a type annotation on both the val and let. We could try to test whether the two types are the same but which notion of sameness should we use ?

**Comment by cpitclaudel** [comment]
>The compiler just generates this warning whenever it encounters a type annotation on both the val and let. 

Not exactly, AFAIU: only the third one of these three examples gets a warning

```
val a : int -> int
let a (x: int) = x + 1

val b : int -> int
let b (x: int) : int = x + 1

val c : int -> int
let c : int -> int = fun x -> x + 1
```

**Comment by kyoDralliam** [comment]
Indeed we could try to just desugar `let c : int -> int = fun x -> x + 1` to `let c = (fun x -> x + 1) <: int -> int` which would correspond to what happens in the first two cases...

**Comment by cpitclaudel** [comment]
@kyoDralliam I think this is fixed, right? :)

**Comment by kyoDralliam** [comment]
Yes I implemented the modification above in 5fa6479aa8118b2fbaaf323b5cee70cc65a7e58a and 434f52943f824fa0ecf3efcfa1ed9e5dd4fb2d1c. (the snapshot is stale but it should be regenerated in a couple of hours).
